src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1448,13 +1448,20 @@ private async Task ProcessCompilationEventsAsync(AnalysisScope analysisScope, bo
                     cancellationToken.ThrowIfCancellationRequested();
 
                     // Kick off tasks to execute syntax tree actions.
-                    var syntaxTreeActionsTask = Task.Run(() => ExecuteSyntaxTreeActions(analysisScope, cancellationToken), cancellationToken);
+                    var syntaxTreeActionsTask = analysisScope.SyntaxTrees.Any()
+                        ? Task.Run(() => ExecuteSyntaxTreeActions(analysisScope, cancellationToken), cancellationToken)
+                        : Task.CompletedTask;
```

Comment:
>: Task.CompletedTask; [](http://example.com/codeflow?start=24&length=21)

Would it make sense to do this in other locations that kick off tasks? Specifically, executeSuppressionActions looks like it's call to getSuppressableDiagnostics could be done before the call to Task.Run and used to filter whether Task.Run needs invoked as it seems to return an empty array most of the time

Reply:
I see over 300 calls to getSuppressableDiagnostics upon solution open, of which 4 return a non-empty array.

Reply:
It looks like `getSuppressableDiagnostics` is a fast call that doesn't need to execute concurrently. I can make that call on the main thread, and then optionally pass that to `ExecuteSuppressionAction` only if needed.

Reply:
That call is currently happening on the calling thread, right? Seems like we wouldn't even be introducing any work on that thread, just creating fewer tasks.